src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const setOfStates = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        setOfStates.push(currentState);
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        setOfStates.push(currentState);
+        break;
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
+        setOfStates.push(currentState);
+        break;
+    }
```

Comment:
Consider adding a default case to the switch statement for error handling. This is a good practice to catch any unexpected or invalid action types.